lightning-invoice/src/payment.rs

```diff
@@ -559,31 +769,38 @@ where
 
 				if payment_id.is_none() {
 					log_trace!(self.logger, "Payment {} has no id; not retrying", log_bytes!(payment_hash.0));
+					self.remove_path_inflight_htlcs(*payment_hash, path);
 				} else if *rejected_by_dest {
 					log_trace!(self.logger, "Payment {} rejected by destination; not retrying", log_bytes!(payment_hash.0));
 					self.payer.abandon_payment(payment_id.unwrap());
+					self.remove_path_inflight_htlcs(*payment_hash, path);
 				} else if retry.is_none() {
 					log_trace!(self.logger, "Payment {} missing retry params; not retrying", log_bytes!(payment_hash.0));
 					self.payer.abandon_payment(payment_id.unwrap());
+					self.remove_path_inflight_htlcs(*payment_hash, path);
 				} else if self.retry_payment(payment_id.unwrap(), *payment_hash, retry.as_ref().unwrap()).is_ok() {
 					// We retried at least somewhat, don't provide the PaymentPathFailed event to the user.
 					return;
```

Comment:
Don't we need to remove it here? The retry could be on an entirely different path or multiple paths.

Reply:
hm, should we not, in this case, rely on the payment retry to send us another event (either `PaymentPathFailed` or `PaymentPathSuccessful`) which we'll then use instead to know when to remove?

Reply:
Yes, we would indeed. But this *is* the event handler, so it needs to adjust for the path that has just failed (i.e., the one it just retried). And, actually, it should really adjust it before calling `retry_payment` otherwise the HTLC that just failed would still be accounted for as in-flight. I imagine we can just call `remove_path_inflight_htlcs` in one place prior to this long `if-else-if` expression.

Reply:
gotcha, thanks!